Add section on integer-based continuous aggregates (#325)
timescale/docs.timescale.com-content: using-timescaledb/continuous-aggregates.md

additionally we require that no `FILTER` clauses appear ([though we will allow them in future
releases](#future-work)).

> ### Integer-based Continuous Aggregates [](create-integer)

Reviewer: ```suggestion
###Continuous Aggregates for integer-based hypertables [](create-integer)
```

> ### Integer-based Continuous Aggregates [](create-integer)

Reviewer: ```suggestion
### Continuous Aggregates for integer-based hypertables [](create-integer)
```

Author: Done.

> Usually, continuous aggregates are defined on a time column, but it is

Reviewer: Great example! The only suggestion is to move this under a new section of the tutorial like Advanced Topics/Advanced Users (or something similar).
This is fairly self-contained and is of interest only to certain users. Also, will not affect the current flow of the tutorial.  Maybe Best Practices and this section go under Advanced Topics?

I am not able to add a comment to unchanged parts of the doc. I see a section called Data retention for continuous aggregates. The paragraph about down sampling and ignore_invalidation_older_than (around line 281) could probably be moved to the Data retention section. I think it is a better fit there.

Author: > I am not able to add a comment to unchanged parts of the doc. I see a section called Data retention for continuous aggregates. The paragraph about down sampling and ignore_invalidation_older_than (around line 281) could probably be moved to the Data retention section. I think it is a better fit there.

The PR for the additions to the data retention section is #314.

Author: > Great example! The only suggestion is to move this under a new section of the tutorial like Advanced Topics/Advanced Users (or something similar).
> This is fairly self-contained and is of interest only to certain users. Also, will not affect the current flow of the tutorial. Maybe Best Practices and this section go under Advanced Topics?

Good idea. I'll add an "Advanced Topics" section. Note that this is not the tutorial, this is the "using" section of the documentation.

Author: Moved section to new "Advanced Topics" section.

  PRIMARY KEY (time)
);

SELECT * FROM create_hypertable('devices', 'time',
                                chunk_time_interval => 1000);
```

To define a continuous aggregate on a table that is using an integer
dimension, it is necessary to have a function to get the current
timestamp in whatever representation that you are using and set it for
the hypertable using
[`set_integer_now_func`][api-set-integer-now-func]. The function can
be defined as a normal Postgres function, but need to be
[`STABLE`][pg-func-stable], take no arguments, and should return an
integer of the same type as the time column in the table. In our case,
this should suffice:

```sql
CREATE OR REPLACE FUNCTION current_microfortnight() RETURNS BIGINT
LANGUAGE SQL STABLE AS $$
	SELECT CAST(1209600 * EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) / 1000000 AS BIGINT)
$$;

SELECT set_integer_now_func('devices', 'current_microfortnight');
```

Once the replacement for current timestamp has been set up, you can
define the continuous aggregate for the `devices` table.

```sql
CREATE VIEW devices_summary
WITH (
  timescaledb.continuous,
  timescaledb.refresh_interval='1 minutes',
  timescaledb.refresh_lag = '-5')

> AS SELECT time_bucket('5', time) AS bucket,

Author: @gayyappan One thing that is not entirely clear here is how `refresh_lag` is used together with the integer-now function, and I think this needs to be added. For inserted rows, are they added when

1. `row.time <= current_fortnight() + refresh_lag`
2. `row.time <= current_fortnight() - refresh_lag`.

The `refresh_lag` is negative here (it was copied from the test), which suggest 1, but usual use of refresh lag suggest 2.

Author: It appears to be 1, from `api.md` at least.

      5 | 69.8000000000000000 |   9.6000000000000000
     10 | 70.8000000000000000 |  24.0000000000000000
     15 | 75.8000000000000000 |  37.6000000000000000
     20 | 71.6000000000000000 |  26.8000000000000000
     25 | 67.6000000000000000 |  56.0000000000000000
     30 | 68.8000000000000000 |  90.2000000000000000
     35 | 71.6000000000000000 |  88.8000000000000000
     40 | 66.4000000000000000 |  81.2000000000000000
     45 | 68.2000000000000000 | 106.0000000000000000
     50 | 72.2000000000000000 | 105.0000000000000000
     55 | 66.2000000000000000 | 138.8000000000000000
     60 | 61.6000000000000000 | 131.4000000000000000
     65 | 64.2000000000000000 | 136.4000000000000000
     70 | 72.0000000000000000 | 114.8000000000000000
     75 | 64.4000000000000000 | 155.2000000000000000
     80 | 65.8000000000000000 | 175.2000000000000000
     85 | 68.0000000000000000 | 156.8000000000000000
     90 | 71.6000000000000000 | 151.4000000000000000
     95 | 68.6000000000000000 | 136.2000000000000000
(20 rows)
```

If you want to refresh the view immediately, you can do it in the
normal manner:

```sql
REFRESH MATERIALIZED VIEW devices_summary;

> ```

Author: @gayyappan When I do this, I get the following errors:

```
postgres=# refresh materialized view devices_summary;
INFO:  new materialization range for public.devices (time column time) larger than allowed in one run, truncating 1915095555 to 100
INFO:  materializing continuous aggregate public.devices_summary: nothing to invalidate, new range up to 100
WARNING:  REFRESH did not materialize the entire range since it was limited by the max_interval_per_job setting
REFRESH MATERIALIZED VIEW
```

Which seems mostly confusing and users are guaranteed to react to it and wonder what they have done wrong. Shall we add a notice to the documentation and/or change the error message?

You can read more about data retention with continuous aggregates in
the [*Data retention*][sec-data-retention] section.

> >:TIP: Most times the continuous aggregate view will be updated by the background job; however, if you would like verify that the view is working correctly, you may use the [`REFRESH MATERIALIZED VIEW` command][api-refresh-continuous-aggs].

Reviewer 2: This is a bit misleading IMO. The background job to refresh the view can be failing even if a manual refresh updates the view. So, I don't think a manual refresh is good  indication of the automation working.

Author: Removed the tip.

> >:TIP: Most times the continuous aggregate view will be updated by the background job; however, if you would like verify that the view is working correctly, you may use the [`REFRESH MATERIALIZED VIEW` command][api-refresh-continuous-aggs].

Reviewer 3: The rest of PR uses multiple lines if sentences or paragraphs are too long, which is nice. Can it be fixed to into multilines here too, so it will be as originally and possible to track changes?

Author: I removed this paragraph for the time being since we consider `REFRESH MATERIALIZED VIEW` a debug command, but not sure what alternative we have for testing the view right now, so it might re-surface.

> >:TIP: Most times the continuous aggregate view will be updated by the background job; however, if you would like verify that the view is working correctly, you may use the [`REFRESH MATERIALIZED VIEW` command][api-refresh-continuous-aggs].

Reviewer 3: I don't understand how calling the refresh will verify that cagg bgw works correctly. Thus I feel the current formulation is misleading.

Author: Because running `REFRESH MATERIALIZED VIEW` will dispatch a job to update the view, so it will indirectly check that the view works. Several users have stumbled over that they set up the view and then don't see it populated, so we have to direct them to use `REFRESH MATERIALIZED VIEW`.

of time that can be represented as integer values, such as nanosecond
epochs, minutes since founding date, or whatever is suitable for your
application.

As an example, suppose that you have a table with CPU and disk usage
for some devices where time is measured in
[microfortnights][fff-system] (a microfortnight is a little more than
a second). Since you are using an integer-valued column as time, you
need to provide the chunk time interval when creating the
hypertable. In this case, let each chunk consist of a millifortnight
(a 1000 microfortnights, which is about 20 minutes).

```sql
CREATE TABLE devices(
  time BIGINT,        -- Timestamp in microfortnights
  cpu_usage INTEGER,  -- Total CPU usage
  disk_usage INTEGER, -- Total disk usage
  PRIMARY KEY (time)
);

SELECT create_hypertable('devices', 'time',
                         chunk_time_interval => 1000);
```

To define a continuous aggregate on a hypertable that is using an integer
dimension, it is necessary to have a function to get the current
timestamp in whatever representation that you are using and set it for
the hypertable using
[`set_integer_now_func`][api-set-integer-now-func]. The function can

> be defined as a normal PostgreSQL function, but needs to: be

Reviewer 3: Typo:
```suggestion
be defined as a normal PostgreSQL function, but needs to be
```

---

### Advanced Topics [](advanced-usage)

#### Continuous Aggregates using Integer-Based Time [](create-integer)

Usually, continuous aggregates are defined on a [date/time-type](https://www.postgresql.org/docs/current/datatype-datetime.html) column, but it is
also possible to create your own custom scheme for handling
aggregation for tables that are using an integer time
column. This can be useful if you have tables that use other measures
of time that can be represented as integer values, such as nanosecond
epochs, minutes since founding date, or whatever is suitable for your
application.

As an example, suppose that you have a table with CPU and disk usage
for some devices where time is measured in
[microfortnights][fff-system]. Since you are using an integer-valued
column as time, you need to provide the chunk time interval when

> creating the hypertable. In this case, let each chunk consist of a

Reviewer 3: I guess you meant 1000 millifortnights, otherwise the sentence is repeating the first sentence:

```suggestion
creating the hypertable. In this case, let each chunk consist of 1000
```

> millifortnight (which is about 20 minutes).

Reviewer 3: Then it will be:
```suggestion
millifortnights and each millifortnight is about 20 minutes.
```

Author: The chunks are one millifortnight, not 1000.

```sql
CREATE TABLE devices(
  time BIGINT,        -- Timestamp in microfortnights
  cpu_usage INTEGER,  -- Total CPU usage
  disk_usage INTEGER, -- Total disk usage
  PRIMARY KEY (time)
);

SELECT create_hypertable('devices', 'time',
                         chunk_time_interval => 1000);
```

> To define a continuous aggregate on a table that is using an integer

Reviewer 2: ```suggestion
To define a continuous aggregate on a hypertable that is using an integer
```

> dimension, it is necessary to have a function to get the current

Reviewer 3: I think it is good to mention that it is `time dimension` and `integer` is a type of it.
```suggestion
To define a continuous aggregate on a table that is using an integer
time dimension, it is necessary to have a function to get the current
```
or 
```suggestion
To define a continuous aggregate on a table that is using an integer for
the time dimension, it is necessary to have a function to get the current
```

Author: Done.

> dimension, it is necessary to have a function to get the current
> timestamp in whatever representation that you are using and set it for

Reviewer 3: It might be just personal opinion:
```suggestion
dimension, it is necessary to have a function to get the current
time in whatever representation that you are using and set it for
```

Author: I'm talking about timestamp everywhere else and am using the `CURRENT_TIMESTAMP` to get the epoch, so I think that "timestamp" is more consistent.

Reviewer 3: Section `Using Continuous Aggregates` in this file uses `current time`, so `current timestamp` is not consistent with the rest of the document. I don't see that you use `timestamp` in the text, except here. `current time` is easier to follow and will be consistent with the rest of the document.

Author: Good point. Switched to use `current_time` and removing uses of "timestamp" in this commit. There are still some mention of timestamps in the chapter, but that is for a separate PR.

the hypertable using
[`set_integer_now_func`][api-set-integer-now-func]. The function can
be defined as a normal PostgreSQL function, but needs to: be

> [`STABLE`][pg-func-stable], take no arguments, and return an integer
> of the same type as the time column in the table. In our case, this

Reviewer 3: ```suggestion
[`STABLE`][pg-func-stable], take no arguments, and return an integer value
of the same type as the time column in the table. In our case, this
```
Or just value.

> should suffice:

Reviewer 3: Will it be better to move the fact that millifortnight is about 20 mins here, since this fact is used in calculating current time?

```sql
CREATE OR REPLACE FUNCTION current_microfortnight() RETURNS BIGINT
LANGUAGE SQL STABLE AS $$
	SELECT CAST(1209600 * EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) / 1000000 AS BIGINT)
$$;

SELECT set_integer_now_func('devices', 'current_microfortnight');
```

Once the replacement for current timestamp has been set up, you can

> define the continuous aggregate for the `devices` table.

Reviewer 3: I guess it should be `a` continuous aggregate, since it (its shape) wasn't described before.
```suggestion
define a continuous aggregate for the `devices` table.
```

```sql
CREATE VIEW devices_summary
WITH (timescaledb.continuous) AS
SELECT time_bucket('500', time) AS bucket,
       avg(cpu_usage) AS avg_cpu,
       avg(disk_usage) AS avg_disk
     FROM devices
     GROUP BY bucket;
```

You can now insert some rows to check if the aggregation works as
expected.

```sql
CREATE EXTENSION tablefunc;

INSERT INTO devices(time, cpu_usage, disk_usage)
SELECT time,
       normal_rand(1,70,10) AS cpu_usage,
	   normal_rand(1,2,1) * (row_number() over()) AS disk_usage
  FROM generate_series(1,10000) AS time;
```

> >:TIP: You can use the `tablefunc` extension to generate a Gaussian
> >distribution and use the `row_number` function to turn it into a

Reviewer 3: I believe `normal distribution` is more commonly used and it matches the name of SQL function, which will be useful for those users, who are not familiar with statistics.
```suggestion
>:TIP: You can use the `tablefunc` extension to generate a normal
>distribution and use the `row_number` function to turn it into a
```

>cumulative sequence.

You can now check that the view contains the correct data. 

```sql
postgres=# SELECT * FROM devices_summary ORDER BY bucket LIMIT 10;
 bucket |       avg_cpu       |       avg_disk
--------+---------------------+----------------------
      0 | 63.0000000000000000 |   6.0000000000000000
      5 | 69.8000000000000000 |   9.6000000000000000
     10 | 70.8000000000000000 |  24.0000000000000000
     15 | 75.8000000000000000 |  37.6000000000000000
     20 | 71.6000000000000000 |  26.8000000000000000
     25 | 67.6000000000000000 |  56.0000000000000000
     30 | 68.8000000000000000 |  90.2000000000000000
     35 | 71.6000000000000000 |  88.8000000000000000
     40 | 66.4000000000000000 |  81.2000000000000000
     45 | 68.2000000000000000 | 106.0000000000000000
(10 rows)
```

Note that the continuous aggregate will lag behind the hypertable 500
microfortnights (since the bucket size is 500 microfortnights) so you

> have to wait for it to be updated.

Reviewer 3: Why? It is the doc for 1.7, where real time aggregates are added. Should the query produce the complete result?

---

### Advanced Topics [](advanced-usage)

#### Continuous Aggregates using Integer-Based Time [](create-integer)

> Usually, continuous aggregates are defined on a [date/time-type](https://www.postgresql.org/docs/current/datatype-datetime.html) column, but it is

Reviewer 4: Can you make this into a footnote-style reference please

Author: Done.

>distribution and use the `row_number` function to turn it into a
>cumulative sequence.

You can now check that the view contains the correct data. 

```sql
postgres=# SELECT * FROM devices_summary ORDER BY bucket LIMIT 10;
 bucket |       avg_cpu       |       avg_disk
--------+---------------------+----------------------
      0 | 63.0000000000000000 |   6.0000000000000000
      5 | 69.8000000000000000 |   9.6000000000000000
     10 | 70.8000000000000000 |  24.0000000000000000
     15 | 75.8000000000000000 |  37.6000000000000000
     20 | 71.6000000000000000 |  26.8000000000000000
     25 | 67.6000000000000000 |  56.0000000000000000
     30 | 68.8000000000000000 |  90.2000000000000000
     35 | 71.6000000000000000 |  88.8000000000000000
     40 | 66.4000000000000000 |  81.2000000000000000
     45 | 68.2000000000000000 | 106.0000000000000000
(10 rows)
```

Note that the continuous aggregate will lag behind the hypertable 500
microfortnights (since the bucket size is 500 microfortnights) so you
have to wait for it to be updated. If you want to refresh the
continuous aggregate view immediately, you can use [`REFRESH
MATERIALIZED VIEW`][api-refresh-continuous-aggs]:

```sql

> REFRESH MATERIALIZED VIEW devices_summary;

Reviewer: There are still references to REFRESH MATERIALIZED VIEW on lines 350-355.

Author: Still not sure how to handle this. We are considering `REFRESH MATERIALIZED VIEW` as a debug command, but users will run into the issue that the materialization has not executed, so how should we direct them to having it updated?

Author: I removed this for now, as a separate commit, so that I can restore it if necessary.

---

### Advanced Topics [](using-cagg-advanced)

> #### Continuous Aggregates for Hypertables using Integer-Based Time [](create-integer)

Reviewer 4: ```suggestion
#### Continuous Aggregates using Integer-Based Time [](create-integer)
```
'Hypertables' is implied here?

of time that can be represented as integer values, such as nanosecond
epochs, minutes since founding date, or whatever is suitable for your
application.

As an example, suppose that you have a table with CPU and disk usage
for some devices where time is measured in
[microfortnights][fff-system]. Since we are using an integer-valued
column as time, we need to provide the chunk time interval when
creating the hypertable. In this case, we let each chunk consist of a
millifortnight.

```sql
CREATE TABLE devices(
  time BIGINT,        -- Timestamp in microfortnights
  cpu_usage INTEGER,  -- Total CPU usage
  disk_usage INTEGER, -- Total disk usage
  PRIMARY KEY (time)
);

SELECT * FROM create_hypertable('devices', 'time',
                                chunk_time_interval => 1000);
```

To define a continuous aggregate on a table that is using an integer
dimension, it is necessary to have a function to get the current
timestamp in whatever representation that you are using and set it for
the hypertable using
[`set_integer_now_func`][api-set-integer-now-func]. The function can
be defined as a normal PostgreSQL function, but need to be

> [`STABLE`][pg-func-stable], take no arguments, and should return an

Reviewer 4: ```suggestion
be defined as a normal PostgreSQL function, but needs to: be
[`STABLE`][pg-func-stable], take no arguments, and return an
```

integer of the same type as the time column in the table. In our case,
this should suffice:

```sql
CREATE OR REPLACE FUNCTION current_microfortnight() RETURNS BIGINT
LANGUAGE SQL STABLE AS $$
	SELECT CAST(1209600 * EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) / 1000000 AS BIGINT)
$$;

SELECT set_integer_now_func('devices', 'current_microfortnight');
```

Once the replacement for current timestamp has been set up, you can
define the continuous aggregate for the `devices` table.

```sql
CREATE VIEW devices_summary
WITH (timescaledb.continuous,
      timescaledb.refresh_interval = '1 minute',
      timescaledb.refresh_lag = '-500') AS
SELECT time_bucket('500', time) AS bucket,
       avg(cpu_usage) AS avg_cpu,
       avg(disk_usage) AS avg_disk
     FROM devices

>     GROUP BY bucket;

Reviewer 4: ```suggestion
     GROUP BY bucket;
```

---

### Advanced Topics [](using-cagg-advanced)

> #### Continuous Aggregates for Integer-Based Hypertables [](create-integer)

Reviewer 2: ```suggestion
#### Continuous Aggregates for Hypertables using Integer-Based Time [](create-integer)
```

> Usually, continuous aggregates are defined on a time column, but it is

Reviewer 2: ```suggestion
Usually, continuous aggregates are defined on a [date/time-type](https://www.postgresql.org/docs/current/datatype-datetime.html) column, but it is
```

also possible to create your own custom scheme for handling
aggregation for tables that are using an integer for the time

> column. This can be useful if you have tables that use other measures

Reviewer 2: ```suggestion
aggregation for tables that are using an integer time
column. This can be useful if you have tables that use other measures
```

of time that can be represented as integer values, such as nanosecond
epochs, minutes since founding date, or whatever is suitable for your
application.

As an example, suppose that you have a table with CPU and disk usage
for some devices where time is measured in
[microfortnights][fff-system]. Since we are using an integer-valued
column as time, we need to provide the chunk time interval when
creating the hypertable. In this case, we let each chunk consist of a
millifortnight.

```sql
CREATE TABLE devices(
  time BIGINT,        -- Timestamp in microfortnights
  cpu_usage INTEGER,  -- Total CPU usage
  disk_usage INTEGER, -- Total disk usage
  PRIMARY KEY (time)
);

SELECT * FROM create_hypertable('devices', 'time',
                                chunk_time_interval => 1000);
```

To define a continuous aggregate on a table that is using an integer
dimension, it is necessary to have a function to get the current
timestamp in whatever representation that you are using and set it for
the hypertable using
[`set_integer_now_func`][api-set-integer-now-func]. The function can

> be defined as a normal Postgres function, but need to be

Reviewer 2: ```suggestion
be defined as a normal PostgreSQL function, but need to be
```

[`STABLE`][pg-func-stable], take no arguments, and should return an
integer of the same type as the time column in the table. In our case,
this should suffice:

```sql
CREATE OR REPLACE FUNCTION current_microfortnight() RETURNS BIGINT
LANGUAGE SQL STABLE AS $$
	SELECT CAST(1209600 * EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) / 1000000 AS BIGINT)
$$;

SELECT set_integer_now_func('devices', 'current_microfortnight');
```

Once the replacement for current timestamp has been set up, you can
define the continuous aggregate for the `devices` table.

```sql
CREATE VIEW devices_summary
WITH (timescaledb.continuous,
      timescaledb.refresh_interval = '1 minute',

>       timescaledb.refresh_lag = '-500') AS

Reviewer 2: Unclear why we are using this `refresh_lag` setting. Doesn't this make the aggregate very inefficient (refresh always)?

Just worried that users are following this example because they don't understand these settings (I am myself confused by them) and then don't get the performance they expect.

It is also confusing that the `refresh_interval` and `refresh_lag` are using different time representations.

Author: If we do not set the `refresh_lag` the view will not be updated, which might give the appearance that the view is not working. A refresh lag set to this value just means that the view is updated for recent values, so AFAICT that would mean that the "most recent" value might not be accurate, but not affect performance.

Than the two parameters use different time representations does not seem strange to me. The refresh interval gives how often the refresh is executed and should be given in wall time, but the `refresh_lag` is referring to microfortnights in this example, so it makes sense that it is an integer.

Reviewer 2: In other places in our docs, we state there are caveats with setting the refresh_lag like this and it is not the default setting. For instance, our docs says: `Tuning the refresh_lag parameter lower will mean that the aggregates will follow inserts more closely, but can cause some write amplification which may degrade insert performance. `. And, further, the example setting here is not only a "magic number" but also negative!, which I think is confusing for the reader. So, in this example we're basically not only ignoring our own best practices, but we're also confusing the user with very non-intuitive settings.

With default settings, isn't it only the most recent bucket that is not updated? Can't we handle that by explaining in the tutorial that this is how continuous aggs work?

I think it is better to explain the recommended behavior than trying to circumvent it ourselves.

SELECT time_bucket('500', time) AS bucket,
       avg(cpu_usage) AS avg_cpu,
       avg(disk_usage) AS avg_disk
     FROM devices
    GROUP BY bucket;
```

You can now insert some rows to check if the aggregation works as
expected. In this case, we are using the `tablefunc` extension to
generate a Gaussian distribution of CPU usage and a cumulative growth
of disk usage.

```sql
CREATE EXTENSION tablefunc;

INSERT INTO devices(time, cpu_usage, disk_usage)
SELECT time,
       normal_rand(1,70,10) AS cpu_usage,
	   normal_rand(1,2,1) * (row_number() over()) AS disk_usage
  FROM generate_series(1,10000) AS time;
```

> You can now check that the view contain the correct data.

Reviewer 2: ```suggestion
You can now check that the view contains the correct data.
```

```sql

> postgres=# SELECT * FROM devices_summary ORDER BY bucket;

Reviewer 2: Maybe trim the output, LIMIT to 10 rows or something.

Author: Done.

      0 | 63.0000000000000000 |   6.0000000000000000
      5 | 69.8000000000000000 |   9.6000000000000000
     10 | 70.8000000000000000 |  24.0000000000000000
     15 | 75.8000000000000000 |  37.6000000000000000
     20 | 71.6000000000000000 |  26.8000000000000000
     25 | 67.6000000000000000 |  56.0000000000000000
     30 | 68.8000000000000000 |  90.2000000000000000
     35 | 71.6000000000000000 |  88.8000000000000000
     40 | 66.4000000000000000 |  81.2000000000000000
     45 | 68.2000000000000000 | 106.0000000000000000
     50 | 72.2000000000000000 | 105.0000000000000000
     55 | 66.2000000000000000 | 138.8000000000000000
     60 | 61.6000000000000000 | 131.4000000000000000
     65 | 64.2000000000000000 | 136.4000000000000000
     70 | 72.0000000000000000 | 114.8000000000000000
     75 | 64.4000000000000000 | 155.2000000000000000
     80 | 65.8000000000000000 | 175.2000000000000000
     85 | 68.0000000000000000 | 156.8000000000000000
     90 | 71.6000000000000000 | 151.4000000000000000
     95 | 68.6000000000000000 | 136.2000000000000000
(20 rows)
```

If you want to refresh the view immediately, you can do it in the
normal manner:

```sql

> REFRESH MATERIALIZED VIEW devices_summary;

Reviewer 2: Note that this might be confusing because apparently this is designed to not be called directly. Also, it won't always refresh the view as one would expect (depends on your view settings). I guess it works in this case because you configured a `refresh_lag` of `-bucket`, which is not the recommended (or default) setting since it makes the agg inefficient.

I know this is not intuitive, but that's also why I am wondering if we shouldn't mention manual refresh since it currently doesn't do what you expect in all cases.

Author: It is already in the manual as a tip above, so we should not shy from mentioning it. If it does not behave as expected in all cases, we should document this, or fix the behavior, but I do not think that omitting mentioning it is a helpful for the users.

Also, I do not think that refresh lag affects performance, see comment above, but I might have misunderstood how it works.

Reviewer 2: The docs says that an aggressive refresh_lag affects insert performance and increases write amplification. (See above).

If refresh_lag would not affect performance, then there is no reason for that parameter to exist at all since one would always prefer no lag.

Author: AFAICT, write amplification does not automatically lead to lower performance, but it does increase utilization. However, I still think that we should keep the `REFRESH MATERIALIZED VIEW` given that it is already discussed elsewhere in the documentation.

Reviewer 2: Ok, let's keep `REFRESH MATERIALIZED VIEW`. Fwiw, I think this command _should_ be a first-class command for users since it is for regular views. It's just that I've heard some people say it is not supposed to be used (and it  prints a lot of cryptic info) and semantics of it is currently confusing since it doesn't completely refresh the view like you would expect.

Reviewer: @mkindahl I think this is the REFRESH ... that you referring to:
_:TIP: Most times the continuous aggregate view will be updated by the background job; however, if you would like to run it yourself, you may use the REFRESH MATERIALIZED VIEW command._

should be rephrased as : 
:TIP: The continuous aggregate view will be updated by the background job; however, if you would like to verify if the view is working correctly, you may use the REFRESH MATERIALIZED VIEW command.

Author: Fixed the tip, but if `alter_job_schedule` is going to be used, you need to dig out the `job_id`, which is not obvious how to do, so I would say that we either keep the `REFRESH` even though it is not perfect, or remove it entirely.

I think it is better to keep it since it does ensure that the view shows *something*, even it if is not perfect. The usage in the tutorial is more reminiscent of what you would do when debugging (just see that it works, don't really care about the exact figures).

Reviewer: If you are keeping REFRESH, please add a note that this is generally how we manually check if the view is working and not running into any errors.The REFRESH MATERIALIZED VIEW will behave like  a _single run_ of the (automatically run) background worker job. (I don't think we actually say this anywhere else).
Please modify the verbiage around this paragraph because it could be interpreted as the entire view is getting updated

_If you want to refresh the
continuous aggregate view immediately, you can use [`REFRESH
MATERIALIZED VIEW`][api-refresh-continuous-aggs]:_

Reviewer 2: As per my other comment, I think saying that you can use a manual refresh as a method of checking if things are working is highly misleading. A background job could be failing or missing even if a manual refresh works.

Reviewer: @erminator If the manual refresh works, the background job that does the automatic refresh should work as well.i.e to say if manual refresh does not work => bgw job will fail. (Yes, there are other cases where bgw jobs  could fail but they are related to the job scheduling and not to the materialization itself). We have a materialization bug if the bgw job materialization fails but the refresh succeeds

Author: Removed all text mentioning the `REFRESH MATERIALIZED VIEW`.

---

> [fff-system]: https://dl.acm.org/doi/10.1145/1229899.1229919

Reviewer 4: Not sure this is the correct link?  Led me to:

<img width="988" alt="Screen Shot 2020-03-09 at 3 35 06 PM" src="https://user-images.githubusercontent.com/12288273/76250402-af207380-621b-11ea-8c3b-4cee1c864246.png">

Author: It is actually the right link, but maybe I should pick another one instead that is not a PDF.

Author: Picked the Wikipedia page instead.

Reviewer 2: I am confused by this. What is the purpose of this link?

[api-continuous-aggregate-stats]: /api#timescaledb_information-continuous_aggregate_stats
[api-drop-chunks]: /api#drop_chunks
[api-set-chunk-interval]: /api#set_chunk_time_interval

> [api-set-integer-now]: /api#set_integer_now_func

Reviewer 4: ```suggestion
[api-set-integer-now-func]: /api#set_integer_now_func
```
